skimage/morphology/_util.py

```diff
@@ -51,35 +54,206 @@ def _validate_connectivity(image_dim, connectivity, offset):
-    squared_distances = np.sum((connection_indices - center) ** 2, axis=1)
-    return offsets[np.argsort(squared_distances)]
+    if selem is None:
+        if connectivity is None:
+            connectivity = ndim
+        selem = ndi.generate_binary_structure(ndim, connectivity)
+    else:
+        # Validate custom structured element
+        selem = np.asarray(selem, dtype=np.bool)
+        # Must specify neighbors for all dimensions
+        if selem.ndim != ndim:
+            raise ValueError(
+                "structuring element and image must have the same number of "
+                "dimensions"
+            )
+        # Must only specify direct neighbors
+        if any(s != 3 for s in selem.shape):
+            raise ValueError("dimension size in structuring element is not 3")
+
+    return selem
+
+
+def _set_edge_values_inplace(image, value):
+    """Set edge values along all axes to a constant value.
+
+    Parameters
+    ----------
+    image : ndarray
+        The array to modify inplace.
+    value : scalar
+        The value to use. Should be compatible with `image`'s dtype.
+
+    Examples
+    --------
+    >>> image = np.zeros((4, 5), dtype=int)
+    >>> _set_edge_values_inplace(image, 1)
+    >>> image
+    array([[1, 1, 1, 1, 1],
+           [1, 0, 0, 0, 1],
+           [1, 0, 0, 0, 1],
+           [1, 1, 1, 1, 1]])
+    """
+    for axis in range(image.ndim):
+        # Index first and last element in each dimension
+        sl = (slice(None),) * axis + ((0, -1),) + (...,)
+        image[sl] = value
+
+
+def _fast_pad(image, value):
+    """Pad an array on all axes by one with a value.
+
+    Parameters
+    ----------
+    image : ndarray
+        Image to pad.
+    value : scalar
+         The value to use. Should be compatible with `image`'s dtype.
+
+    Returns
+    -------
+    padded_image : ndarray
+        The new image.
+
+    Notes
+    -----
+    The output of this function is equivalent to::
+
+        np.pad(image, 1, mode="constant", constant_values=value)
+
+    Up to versions < 1.17 `numpy.pad` uses concatenation to create padded
+    arrays while this method needs to only allocate and copy once.
+    This can result in significant speed gains if `image` has a large number of
+    dimensions.
+    Thus this function may be safely removed once that version is the minimum
+    required by scikit-image.
+
+    Examples
+    --------
+    >>> _fast_pad(np.zeros((2, 3), dtype=int), 4)
+    array([[4, 4, 4, 4, 4],
+           [4, 0, 0, 0, 4],
+           [4, 0, 0, 0, 4],
+           [4, 4, 4, 4, 4]])
+    """
+    # Allocate padded image
+    new_shape = np.array(image.shape) + 2
+    new_image = np.empty(new_shape, dtype=image.dtype, order="C")
+
+    # Copy old image into new space
+    sl = (slice(1, -1),) * image.ndim
+    new_image[sl] = image
+    # and set the edge values
+    _set_edge_values_inplace(new_image, value)
+
+    return new_image
```

Comment:
Very nice! =)

Reply:
LOL Just saw that this already existed. :joy:

Reply:
Well I did tweak the way slices are constructed here and `_set_edge_values_inplace`. So it doesn't hurt that you looked at it again with new eyes. :smile: